src/ellswift.rs

```diff
@@ -0,0 +1,705 @@
+        mut hash_function: F,
+    ) -> ElligatorSwiftSharedSecret
+    where
+        F: FnMut([u8; 32], [u8; 64], [u8; 64]) -> ElligatorSwiftSharedSecret,
+    {
+        let mut shared_secret = [0u8; 32];
+        let hashfp = hash_callback::<F>;
+        unsafe {
+            let ret = ffi::secp256k1_ellswift_xdh(
+                ffi::secp256k1_context_no_precomp,
+                shared_secret.as_mut_c_ptr(),
+                ellswift_a.0.as_c_ptr(),
+                ellswift_b.0.as_c_ptr(),
+                secret_key.as_c_ptr(),
+                party.to_ffi_int(),
+                Some(hashfp),
+                &mut hash_function as *mut F as *mut c_void,
+            );
+            debug_assert_eq!(ret, 1);
+        }
+        ElligatorSwiftSharedSecret(shared_secret)
+    }
+
+    /// Encodes a public key into an `ElligatorSwift` encoding
+    fn ellswift_encode(pk: PublicKey) -> ElligatorSwift {
+        let mut ell_out = [0u8; constants::ELLSWIFT_ENCODING_SIZE];
+        unsafe {
+            let ret = ffi::secp256k1_ellswift_encode(
+                ffi::secp256k1_context_no_precomp,
+                ell_out.as_mut_c_ptr(),
+                pk.as_c_ptr(),
+                [0u8; 32].as_ptr(),
+            );
+            debug_assert_eq!(ret, 1);
+        }
+        ElligatorSwift(ffi::ElligatorSwift::from_array(ell_out))
+    }
+
+    /// Decodes an `ElligatorSwift` encoding into a [ffi::PublicKey]. This function is a low
+    /// level function, and will be used in internal functions.
+    pub(crate) fn ellswift_decode(ell: ElligatorSwift) -> ffi::PublicKey {
+        unsafe {
+            let mut pk = ffi::PublicKey::new();
+
+            let ret = ffi::secp256k1_ellswift_decode(
+                ffi::secp256k1_context_no_precomp,
+                pk.as_mut_c_ptr(),
+                ell.0.as_c_ptr(),
+            );
+            debug_assert_eq!(ret, 1);
+            pk
+        }
+    }
+}
+
+/// The result of `ElligatorSwift::shared_secret`, which is a shared secret
+/// computed from the x-only ECDH using both parties' public keys (`ElligatorSwift` encoded) and our own
+/// private key.
+#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct ElligatorSwiftSharedSecret([u8; 32]);
+
+/// Represents which party we are in the ECDH, A is the initiator, B is the responder.
+/// This is important because the hash of the shared secret is different depending on which party
+/// we are. In this context, "we" means the party that is using this library, and possesses the
+/// secret key passed to `ElligatorSwift::shared_secret`.
+#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub enum ElligatorSwiftParty {
+    /// We are the initiator of the ECDH
+    A = 0,
+    /// We are the responder of the ECDH
+    B = 1,
+}
+
+impl ElligatorSwiftParty {
+    fn to_ffi_int(self) -> c_int {
+        match self {
+            ElligatorSwiftParty::A => 0,
+            ElligatorSwiftParty::B => 1,
```

Comment:
We are still assigning values here, did you mean to remove those?

Reply:
I think @apoelstra's [comment](https://github.com/rust-bitcoin/rust-secp256k1/pull/627/files/505b0797f84f51d00cb9d94469f1c886813fcbbc#r1342152101) was about using something externally visible like `From`. We need, somehow, to get a 0/1 to give the C code, the alternative is to remove this at all and ask the user to explicitly pass 0/1. I think this is more ergonomic and less error-prone (users can't pass an invalid value, like 2).

Reply:
These values are hard to obtain (I think you need to cast to get them). I think you should drop them and instead have a dedicated method to extract them.

Reply:
>  I think you should drop them and instead have a dedicated method to extract them.

Do you mean an API call? Currently, I have this 

```rust
impl ElligatorSwiftParty {
    fn to_ffi_int(self) -> c_int {
        match self {
            ElligatorSwiftParty::A => 0,
            ElligatorSwiftParty::B => 1,
        }
    }
}
```

and the [generated code](https://godbolt.org/z/8hKWqb4Y4) is pretty straightforward.

Reply:
Yep, exactly. If we have this function then we don't need to assign values to the enum variants.

Reply:
Right. That's what I'm doing on 9b9090b

```rust
pub enum ElligatorSwiftParty {
    A,
    B,
}

impl ElligatorSwiftParty {
    fn to_ffi_int(self) -> c_int {
        match self {
            ElligatorSwiftParty::A => 0,
            ElligatorSwiftParty::B => 1,
        }
    }
}
```

No enum variant value, but the user exposed API doesn't expose those internal values.

Reply:
Oh I may have caused the confusion, I saw the `to_ffi_int` function but I thought there were explicit values on the enum still. This comment is not showing as "outdated" so I'm guessing I was looking at and old state. My bad.